File: internal/hud/frontend/src/lib/clients/vendorSessions.ts
// Cross-vendor transcript client — GET /api/vendor-sessions[/search].
// Bridges to the agent-context vendor session tools, which read the on-disk
// session transcripts of the vendor CLIs (Claude Code, Codex) on the
// workstation running the daemon. Read-only; `degraded: true` marks the
// bridge-unavailable placeholder, distinct from "no sessions found".

export interface VendorSession {
  vendor: string;
  id: string;
  path: string;
  cwd?: string;
  source?: string;
  /** Human handle: the conversation summary or first user prompt line.
   * Best-effort — absent on older senders/transcripts. */
  title?: string;
  /** Non-interactive transcript tag: "automation" (codex scheduled runs) or
   * "sidechain" (claude subagents + codex worker threads). Absent for a
   * normal user chat. */
  kind?: string;
  started_at?: string;
  modified_at: string;
  size_bytes: number;
  /** Source workstation for federated rows (mirror pushes); absent for
   * transcripts read by this HUD's own bridge. */
  host?: string;
}

export interface VendorSessionMatch {
  vendor: string;
  session_id: string;
  path: string;
  cwd?: string;
  /** 0 when the federating mirror tail-seeked a large transcript and the
   * absolute line number is unknown. */
  line: number;
  role?: string;
  timestamp?: string;
  snippet: string;
  host?: string;
}

export interface VendorSessionListResult {
  sessions: VendorSession[];
  count: number;
  degraded: boolean;
}

export interface VendorSessionSearchResult {
  query: string;
  matches: VendorSessionMatch[];
  count: number;
  degraded: boolean;
}

export interface VendorSessionFilters {
  vendor?: 'claude' | 'codex';
  cwdContains?: string;
  sinceHours?: number;
  limit?: number;
}

export interface VendorSessionSearchFilters extends VendorSessionFilters {
  maxResults?: number;
  maxPerSession?: number;
}

async function parseResponse(res: Response): Promise<any> {
  let data: any = null;
  try {
    data = await res.json();
  } catch {
    data = null;
  }
  if (!res.ok) {
    const msg = data?.error || `${res.status} ${res.statusText}`;
    throw new Error(msg);
  }
  return data;
}

function filterParams(filters: VendorSessionFilters): URLSearchParams {
  const params = new URLSearchParams();
  if (filters.vendor) params.set('vendor', filters.vendor);
  if (filters.cwdContains) params.set('cwd_contains', filters.cwdContains);
  if (filters.sinceHours && filters.sinceHours > 0) params.set('since_hours', String(filters.sinceHours));
  if (filters.limit && filters.limit > 0) params.set('limit', String(filters.limit));
  return params;
}

export async function fetchVendorSessions(
  filters: VendorSessionFilters = {},
): Promise<VendorSessionListResult> {
  const params = filterParams(filters);
  const qs = params.size > 0 ? `?${params.toString()}` : '';
  const res = await globalThis.fetch(`/api/vendor-sessions${qs}`);
  const data = await parseResponse(res);
  return {
    sessions: data?.sessions ?? [],
    count: data?.count ?? 0,
    degraded: data?.degraded ?? false,
  };
}

export async function searchVendorSessions(
  query: string,
  filters: VendorSessionSearchFilters = {},
): Promise<VendorSessionSearchResult> {
  const params = filterParams(filters);
  params.set('query', query);
  if (filters.maxResults && filters.maxResults > 0) params.set('max_results', String(filters.maxResults));
  if (filters.maxPerSession && filters.maxPerSession > 0) {
    params.set('max_per_session', String(filters.maxPerSession));
  }
  const res = await globalThis.fetch(`/api/vendor-sessions/search?${params.toString()}`);
  const data = await parseResponse(res);
  return {
    query: data?.query ?? query,
    matches: data?.matches ?? [],
    count: data?.count ?? 0,
    degraded: data?.degraded ?? false,
  };
}

File: pkg/vendorsessions/vendorsessions.go
// Package vendorsessions reads the on-disk session transcripts written by
// vendor coding agents (Claude Code and Codex) and exposes a unified
// list/search surface over them.
//
// Claude Code stores one JSONL file per session under
// ~/.claude/projects/<flattened-cwd>/<session-uuid>.jsonl; each line carries
// the record's cwd/sessionId/timestamp. Codex stores
// ~/.codex/sessions/YYYY/MM/DD/rollout-<ts>-<uuid>.jsonl whose first line is
// a session_meta record with id/cwd/originator/source.
//
// Neither store is indexed here: listing stats files newest-first and reads
// only the head of each candidate for metadata, and search streams line by
// line under per-file and total caps. This keeps the tools safe to run
// against multi-GB session dirs.
package vendorsessions

import (
	"os"
	"path/filepath"
	"sort"
	"time"
)

// Vendor identifiers accepted by ListOptions.Vendor.
const (
	VendorClaude = "claude"
	VendorCodex  = "codex"
)

// Session kinds beyond a plain interactive chat, derived from transcript
// head metadata. Empty means interactive.
const (
	// KindAutomation marks a Codex scheduled-automation run (session_meta
	// thread_source "automation") — recurring background work that would
	// otherwise be indistinguishable from the user's own chats.
	KindAutomation = "automation"
	// KindSidechain marks a Claude Code subagent transcript (records carry
	// isSidechain: true) — machine-spawned side work, not a user chat.
	KindSidechain = "sidechain"
)

// Session is unified metadata for one vendor session transcript.
type Session struct {
	Vendor string `json:"vendor"`
	ID     string `json:"id"`
	Path   string `json:"path"`
	CWD    string `json:"cwd,omitempty"`
	Source string `json:"source,omitempty"`
	// Title is the human handle for the session: Claude's conversation
	// summary when present, else the opening line of the first real user
	// prompt. Best-effort; empty when the head scan found neither.
	Title string `json:"title,omitempty"`
	// Kind distinguishes non-interactive transcripts ("automation",
	// "sidechain"); empty for a normal user chat.
	Kind       string    `json:"kind,omitempty"`
	StartedAt  time.Time `json:"started_at,omitempty"`
	ModifiedAt time.Time `json:"modified_at"`
	SizeBytes  int64     `json:"size_bytes"`
}

// ListOptions filters session listing.
type ListOptions struct {
	// Vendor restricts to "claude" or "codex"; empty means both.
	Vendor string
	// CwdContains keeps sessions whose working directory contains this
	// substring (case-sensitive path fragment, e.g. "services/loom-core").
	CwdContains string
	// Since keeps sessions modified at or after this time (zero = no floor).
	Since time.Time
	// Limit caps the number of sessions returned (default 50, max 500).
	Limit int
}

// Store locates the vendor session roots. Zero-value roots are filled from
// the current user's home directory.
type Store struct {
	ClaudeRoot string
	CodexRoot  string
}

// DefaultStore resolves the conventional vendor session roots under $HOME.
func DefaultStore() Store {
	home, err := os.UserHomeDir()
	if err != nil {
		return Store{}
	}
	return Store{
		ClaudeRoot: filepath.Join(home, ".claude", "projects"),
		CodexRoot:  filepath.Join(home, ".codex", "sessions"),
	}
}

func (o *ListOptions) normalize() {
	if o.Limit <= 0 {
		o.Limit = 50
	}
	if o.Limit > 500 {
		o.Limit = 500
	}
}

// List returns unified session metadata across the configured vendors,
// newest-modified first. Missing roots are skipped, not errors: an agent
// host that has never run one of the vendors simply contributes nothing.
func (s Store) List(opts ListOptions) ([]Session, error) {
	opts.normalize()

	var sessions []Session
	if opts.Vendor == "" || opts.Vendor == VendorClaude {
		sessions = append(sessions, listClaude(s.ClaudeRoot, opts)...)
	}
	if opts.Vendor == "" || opts.Vendor == VendorCodex {
		sessions = append(sessions, listCodex(s.CodexRoot, opts)...)
	}

	sort.Slice(sessions, func(i, j int) bool {
		return sessions[i].ModifiedAt.After(sessions[j].ModifiedAt)
	})
	if len(sessions) > opts.Limit {
		sessions = sessions[:opts.Limit]
	}
	return sessions, nil
}

// candidateFiles collects transcript file paths with stat info, newest
// first, bounded by mtime floor. maxFiles bounds how many survive.
type candidate struct {
	path    string
	modTime time.Time
	size    int64
}

func sortCandidates(cands []candidate) {
	sort.Slice(cands, func(i, j int) bool {
		return cands[i].modTime.After(cands[j].modTime)
	})
}
